### ISSUE TITLE
[no-unnecessary-type-assertion] False positive with lazily initialized local variable

### ISSUE THREAD

**Issue by fasttime** [typescript-eslint/typescript-eslint#2423]
Not sure if or how this relates to #2248. In my case, the repro is quite simple.

**Repro**

***.eslintrc.json***
```JSON
{
    "parser": "@typescript-eslint/parser",
    "parserOptions": {
        "project": "./tsconfig.json"
    },
    "plugins": [
        "@typescript-eslint"
    ],
    "rules": {
        "@typescript-eslint/no-unnecessary-type-assertion": [
            "error"
        ]
    }
}
```

***index.ts***
```TS
let foo: string[];
Reflect.apply(() => foo = ['bar'], null, []);
console.log(foo![0]);
```

***package.json***
```JSON
{
  "devDependencies": {
    "@typescript-eslint/eslint-plugin": "latest",
    "@typescript-eslint/parser": "latest",
    "eslint": "7.4.0",
    "typescript": "4.0.2"
  }
}
```

***tsconfig.json***
```JSON
{
    "compilerOptions": {
        "strictNullChecks": false
    }
}
```

Run with: `npx eslint index.ts`

**Expected Result**

No error.

**Actual Result**

Getting error `This assertion is unnecessary since it does not change the type of the expression  @typescript-eslint/no-unnecessary-type-assertion` on line 3.

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `3.10.0` |
| `@typescript-eslint/parser`        | `3.10.0` |
| `TypeScript`                       | `4.0.2` |
| `ESLint`                           | `7.4.0` |
| `node`                             | `14.8.0` |

**Comment by bradzacher** [comment]
I am unable to repro this against master.
Are you using `strictNullChecks`?

**Comment by fasttime** [comment]
I am not using `strictNullChecks`, and I am still getting the error after updating `@typescript-eslint/eslint-plugin` and `@typescript-eslint/parser` to version 4.0.1 (I didn't try cloning the repo and building the packages from master).
I updated the repro with complete versions of the relevant files, that should help reproduce the issue in case the previous repro was missing information.
